Initial implementation

cloudspeak opened the pull request:
Working implementation with a Pulumi stack defining a DynamoDB table, a Firehose delivery stream and an S3 bucket.  Changes in the table are pushed into the delivery stream (by way of the Lambda function in `dynamoTriggerLambda/`), which then delivers the data into the bucket.

A few things to note:

* The region is currently set to eu-west-1 in the config file
* I've left all the AWS resource names blank so they just have the default Pulumi naming
* The Dynamo schema just requires an "Id" string field
* If any records fail to be put into Firehose, this is logged
* I presumed that you would rather the data was stored in a typical JSON format rather than the DynamoDB format, which contains type information.  This being the case, please be aware that since Dynamo store numbers as decimal strings rather than floating point numbers, there is a potential for loss of precision when they are represented as JSON.

LouisAmon commented on firehosePolicy.py:
```diff
@@ -0,0 +1,87 @@
+                    outputs[0],
+                    f'{outputs[0]}/*',
+                    "arn:aws:s3:::%FIREHOSE_BUCKET_NAME%",
+                    "arn:aws:s3:::%FIREHOSE_BUCKET_NAME%/*"
```
Shouldn't this variable be interpolated by `deliveryStreamName` ?

LouisAmon replied:
Overall, the use of `outputs` with positional arguments isn't very explicit as you have to scroll up to read which is what.

Isn't there a way to do the same thing with an explicit dictionary ?

cloudspeak replied:
The reason I left the `%FIREHOSE_BUCKET_NAME%` placeholders is that when you create a delivery stream via the web console, it adds these lines in addition to the actual bucket name, so I assumed they had some purpose.  Now I've Googled it however I've discovered that the [documentation says to remove them](https://docs.aws.amazon.com/firehose/latest/dev/create-configure.html) - bit strange really!  I'll get rid of them.

Agreed about the outputs array, I'll find a nicer solution.

LouisAmon commented on dynamoTriggerLambda/dynamoTriggerLambda.py:
```diff
@@ -0,0 +1,66 @@
+        if num.is_integer():
+            return int(num)
+        else:
+            return num
```
Very nice!
👍

LouisAmon commented on __main__.py:
```diff
@@ -0,0 +1,84 @@
+    }],
+    hash_key='Id',
+    billing_mode='PAY_PER_REQUEST',
+    stream_enabled=True,
```
Yup, that's exactly right!

LouisAmon commented on __main__.py:
```diff
@@ -0,0 +1,84 @@
+    handler='dynamoTriggerLambda.handler',
+    code=pulumi.AssetArchive({
+        ".": pulumi.FileArchive("./dynamoTriggerLambda"),
+    }),
```
Aren't you missing a `.py` extension to the module file ?

Also, isn't it too "verbose" to use both `AssetArchive` and `FileArchive` ? (If we can simplify, all the better)

cloudspeak replied:
I used both `AssetArchive` and `FileArchive` because I saw it in one of the documentation examples, but now you mention it, it doesn't seem to be needed!  I think perhaps they did it in case more archives were added to the asset.  I'll simplify it.

The `.py` extension is missing because it is referring to the folder, not the file, to allow the lambda implementation to be split over multiple files in the future.

LouisAmon commented on __main__.py:
```diff
@@ -0,0 +1,84 @@
+    }),
+    environment={
+        "Variables": {
+            "DELIVERY_STREAM_NAME": deliveryStream.name
```
Yup, you nailed it! 🥇

LouisAmon commented on dynamoTriggerLambda/dynamoTriggerLambda.py:
```diff
@@ -0,0 +1,66 @@
+import boto3
+from boto3.dynamodb.types import TypeDeserializer
+
+DELIVERY_STREAM_VAR_NAME = "DELIVERY_STREAM_NAME"
```
Where is this used ? `DELIVERY_STREAM_VAR_NAME ` ?

cloudspeak replied:
Oops - this was me refactoring the env variable name into a constant and then forgetting to use said constant

LouisAmon commented on dynamoTriggerLambda/dynamoTriggerLambda.py:
```diff
@@ -0,0 +1,66 @@
+def handler(event, context):
+    """ This handler consumes DynamoDB INSERT events, converts the records into a more standard
+        JSON representation, and puts the result into a Kinesis Firehose stream.
+    """
```
What happens to UPDATE & DELETE events ?

Can we replicate those ?

cloudspeak replied:
No problem!

cloudspeak commented on .vscode/settings.json:
```diff
@@ -0,0 +1,12 @@
+        "ints",
+        "pulumi"
+    ],
+    "python.pythonPath": "venv\\Scripts\\python.exe"
```
Sure, I'll add that in

cloudspeak commented on __main__.py:
```diff
@@ -0,0 +1,84 @@
+accountId = get_caller_identity().account_id
+
+async def getRegion():
+    return (await get_region()).id
```
Ah yes, that's much better - I was trying that but kept putting `aws:config` and wondered why it wasn't working!

cloudspeak commented:
I've updated the PR with the following changes:
* Fetch region from the config as suggested
* Tidied up the policy document which was using positional array values
* Removed the Firehose placeholders in the policy document
* Removed the `.vscode` settings folder
* Removed the unnecessary `AssetArchive`
* Used the environment variable name constant in the Lambda
* Added support for Insert, Modify and Remove operations in the dynamo trigger.  The output records now specify the operation in an `Operation` field.  For Insert and Modify events there is also a `Data` field with the data, and for Delete events there is a `Key` field with the key of the deleted item.
